test/integration/controller_test.go

```diff
@@ -358,6 +358,171 @@ func TestAsyncProvisionWithMultiplePolls(t *testing.T) {
 	})
 }
 
+// TestServiceInstanceDeleteWithAsyncUpdateInProgress tests that you can delete
+// an instance during an async update.  That is, if you request a delete during
+// an instance update, the instance will be deleted when the update completes
+// regardless of success or failure.
+func TestServiceInstanceDeleteWithAsyncUpdateInProgress(t *testing.T) {
+
+	cases := []struct {
+		name           string
+		updateSucceeds bool
+	}{
+		{
+			name:           "update succeeds",
+			updateSucceeds: true,
+		},
+		{
+			name:           "update fails",
+			updateSucceeds: false,
+		},
+	}
+
+	for _, tc := range cases {
+		tc := tc
+		t.Run(tc.name, func(t *testing.T) {
+			t.Parallel()
+			const NumberOfInProgressResponses = 2
+			numberOfPolls := 0
+
+			ct := controllerTest{
+				t:                            t,
+				broker:                       getTestBroker(),
+				instance:                     getTestInstance(),
+				skipVerifyingInstanceSuccess: false,
+				setup: func(ct *controllerTest) {
+					ct.osbClient.UpdateInstanceReaction.(*fakeosb.UpdateInstanceReaction).Response.Async = true
+					ct.osbClient.PollLastOperationReaction = fakeosb.DynamicPollLastOperationReaction(
+						func(_ *osb.LastOperationRequest) (*osb.LastOperationResponse, error) {
+							numberOfPolls++
+							state := osb.StateInProgress
+							if numberOfPolls > NumberOfInProgressResponses {
+								if tc.updateSucceeds {
+									state = osb.StateSucceeded
+								} else {
+									state = osb.StateFailed
+								}
+							}
+							return &osb.LastOperationResponse{State: state}, nil
+						})
+				},
+			}
+			ct.run(func(ct *controllerTest) {
+
+				if err := util.WaitForInstanceCondition(ct.client, ct.instance.Namespace, ct.instance.Name,
+					v1beta1.ServiceInstanceCondition{
+						Type:   v1beta1.ServiceInstanceConditionReady,
+						Status: v1beta1.ConditionTrue,
+						Reason: "ProvisionedSuccessfully",
+					}); err != nil {
+					t.Fatalf("error waiting for instance to be ready: %v", err)
+				}
+
+				// add a parameter to the instance for the Update()
+				ct.instance.Spec.Parameters = convertParametersIntoRawExtension(t,
+					map[string]interface{}{
+						"param-key": "new-param-value",
+					})
+
+				_, err := ct.client.ServiceInstances(testNamespace).Update(ct.instance)
+				if err != nil {
+					t.Fatalf("error updating instance: %v", err)
+				}
+
+				if err := util.WaitForInstanceCondition(ct.client, ct.instance.Namespace, ct.instance.Name,
+					v1beta1.ServiceInstanceCondition{
+						Type:   v1beta1.ServiceInstanceConditionReady,
+						Status: v1beta1.ConditionFalse,
+						Reason: "UpdatingInstance",
+					}); err != nil {
+					t.Fatalf("error waiting for instance to be updating asynchronously: %v", err)
+				}
+
+				if err := ct.client.ServiceInstances(ct.instance.Namespace).Delete(ct.instance.Name, &metav1.DeleteOptions{}); err != nil {
```

Comment:
I believe there's a race here between the update completing and the deletion timestamp getting set. Instead of basing the response of the last operation on the number of polls that occurred, I think we should use a boolean (`shouldFinishPolling`?) that we toggle from false to true after issuing the delete request.

Reply:
I concur with @kibbles-n-bytes on this. So that it's clear, the same race exists in the other test.